src/soter/openssl/soter_ec_key_3.c

```diff
@@ -0,0 +1,439 @@
+        return true;
+    } else if (strcmp(curve, SN_secp521r1 /* P-521 */) == 0) {
+        return true;
+    } else {
+        return false;
+    }
+}
+
+/* Input size directly since public key type structures may be aligned to word boundary */
+static size_t ec_pub_key_size(const char* curve, bool compressed)
+{
+    if (strcmp(curve, SN_X9_62_prime256v1 /* P-256 */) == 0) {
+        return sizeof(soter_container_hdr_t)
+               + (compressed ? EC_PUB_SIZE(256) : EC_PUB_UNCOMPRESSED_SIZE(256));
+    } else if (strcmp(curve, SN_secp384r1 /* P-384 */) == 0) {
+        return sizeof(soter_container_hdr_t)
+               + (compressed ? EC_PUB_SIZE(384) : EC_PUB_UNCOMPRESSED_SIZE(384));
+    } else if (strcmp(curve, SN_secp521r1 /* P-521 */) == 0) {
+        return sizeof(soter_container_hdr_t)
+               + (compressed ? EC_PUB_SIZE(521) : EC_PUB_UNCOMPRESSED_SIZE(521));
+    } else {
+        return 0;
+    }
+}
+
+static size_t ec_priv_key_size(const char* curve)
+{
+    if (strcmp(curve, SN_X9_62_prime256v1 /* P-256 */) == 0) {
+        return sizeof(soter_ec_priv_key_256_t);
+    } else if (strcmp(curve, SN_secp384r1 /* P-384 */) == 0) {
+        return sizeof(soter_ec_priv_key_384_t);
+    } else if (strcmp(curve, SN_secp521r1 /* P-521 */) == 0) {
+        return sizeof(soter_ec_priv_key_521_t);
+    } else {
+        return 0;
+    }
+}
+
+static char* ec_pub_key_tag(const char* curve)
+{
+    if (strcmp(curve, SN_X9_62_prime256v1 /* P-256 */) == 0) {
+        return EC_PUB_KEY_TAG(256);
+    } else if (strcmp(curve, SN_secp384r1 /* P-384 */) == 0) {
+        return EC_PUB_KEY_TAG(384);
+    } else if (strcmp(curve, SN_secp521r1 /* P-521 */) == 0) {
+        return EC_PUB_KEY_TAG(521);
+    } else {
+        return NULL;
+    }
+}
+
+static char* ec_priv_key_tag(const char* curve)
+{
+    if (strcmp(curve, SN_X9_62_prime256v1 /* P-256 */) == 0) {
+        return EC_PRIV_KEY_TAG(256);
+    } else if (strcmp(curve, SN_secp384r1 /* P-384 */) == 0) {
+        return EC_PRIV_KEY_TAG(384);
+    } else if (strcmp(curve, SN_secp521r1 /* P-521 */) == 0) {
+        return EC_PRIV_KEY_TAG(521);
+    } else {
+        return NULL;
+    }
+}
+
+static size_t bn_encode(const BIGNUM* bn, uint8_t* buffer, size_t length)
+{
+    int bn_size = BN_num_bytes(bn);
+    if (length < (size_t)bn_size) {
+        return 0;
+    }
+    memset(buffer, 0, length - bn_size);
+    return (length - bn_size) + BN_bn2bin(bn, buffer + (length - bn_size));
+}
+
+soter_status_t soter_engine_specific_to_ec_pub_key(const soter_engine_specific_ec_key_t* engine_key,
+                                                   bool compressed,
+                                                   soter_container_hdr_t* key,
+                                                   size_t* key_length)
+{
+    EVP_PKEY* pkey = (EVP_PKEY*)engine_key;
+    soter_status_t res;
+    size_t output_length;
+    char curve_str[16];
```

Comment:
can you add comment why we declare size of 16? it is curve name with max length? then put name of the reason why we declare 16.

Reply:
Done

Reply:
Actually, `11` was enough here, for longest possible string, `prime256v1\0`, but put `16` because... well, it's a cute power of two

Reply:
your comment doesn't explain why we use 16 bytes size instead of max of required. Let's use expected and understandable value to not force readers to investigate why we used here 16 instead of X when someone will want to increase/decrease this value (for example if somebody decides to add support of another curve with longer name)

Reply:
This is the kind of C coding practice that OpenSSL 3 wants to discourage, btw. The new blessed way is to use the `out_len` parameter to learn the required buffer size, allocate it dynamically, and use that.

I don't have a good answer for this. The most paranoid one would be to do
```c
#define CURVE_P256 SN_X9_62_prime256v1
#define CURVE_P384 SN_secp384r1
#define CURVE_P521 SN_secp521r1
#define CURVE_MAX_SIZE 16
```
then meticulously use our own constants (and drop all those comments explaining that SN_X9_62_prime256v1 is _the_ P-256 in Themis context). And have a test that verifies that `strlen(CURVE_P521) < CURVE_MAX_SIZE`.